src/mpid/ch4/generic/am/mpidig_am_msg.h

```diff
@@ -77,13 +77,29 @@ MPL_STATIC_INLINE_PREFIX void MPIDIG_recv_copy(void *in_data, MPIR_Request * rre
         /* otherwise if recv size = 0, it is at least a truncation error */
         MPIR_STATUS_SET_COUNT(rreq->status, 0);
     } else if (p->recv_type == MPIDIG_RECV_DATATYPE) {
+        MPI_Aint max_data_size;
+        MPIR_Datatype_get_size_macro(MPIDIG_REQUEST(rreq, datatype), max_data_size);
+        max_data_size *= MPIDIG_REQUEST(rreq, count);
+
         MPI_Aint actual_unpack_bytes;
         MPIR_Typerep_unpack(in_data, in_data_sz,
                             MPIDIG_REQUEST(rreq, buffer),
                             MPIDIG_REQUEST(rreq, count),
                             MPIDIG_REQUEST(rreq, datatype), 0, &actual_unpack_bytes);
-        if (in_data_sz > actual_unpack_bytes) {
+        if (in_data_sz > max_data_size) {
+            /* if we received more data than what the receive buffer
+             * can accommodate, it's a truncation error */
             rreq->status.MPI_ERROR = MPIDIG_ERR_TRUNCATE(actual_unpack_bytes, in_data_sz);
+        } else if (in_data_sz > actual_unpack_bytes) {
+            /* If the receive buffer had enough space, but we still
+             * couldn't unpack the data, it means that the basic
+             * datatype from the sender doesn't match that of the
+             * receiver.  This doesn't catch all errors; for example,
+             * cases where the sizes match but the basic datatypes are
+             * still different. */
```

Comment:
Thanks for adding the comments. It is too verbose. What I am looking for is to explain what does `in_data_sz > actual_unpack_bytes` mean? I was confused because it doesn't really mean `dtypemismatch`. It is due to underlying basic element size mismatch. I guess if you want it mean the basic type mismatch, that's fine, but when we review both code together, we'll check whether `MPIR_Type_unpack` will return less `actual_unpack_bytes` when that happen or we need fix code there. Whatever we comment here and the actual code in the `Typerep` implementation need be consistent.

Reply:
`dtypemismatch` is a signature mismatch in all of the MPICH code.  We never expect the actual datatypes to be the same anywhere.  I'm not going to reduce the comment just because it's verbose.

Reply:
I am not complaining on the error message. I was looking for the meaning/understanding of `in_data_sz > actual_unpak_bytes`, not from standard point of view, but from code point of view -- under what condition the code packs less data than expected. I felt the comment not quite explains it. For example, if I send `MPI_INT` to `MPI_FLOAT`, it won't cause this condition, but the comment says it will.

Reply:
It already says it:

```
This doesn't catch all errors; for example,
+             * cases where the sizes match but the basic datatypes are
+             * still different.
```

Reply:
Sorry, I see you added a sentence to address that :)